packages/@aws-cdk/toolkit-lib/lib/api/cloud-assembly/private/source-builder.ts

```diff
@@ -219,3 +259,53 @@ export abstract class CloudAssemblySourceBuilder {
 function noUndefined<A>(xs: Record<string, A>): Record<string, NonNullable<A>> {
   return Object.fromEntries(Object.entries(xs).filter(([_, v]) => v !== undefined)) as any;
 }
+
+/**
+ * Turn synthesis options into context/environment variables that will go to the CDK app
+ *
+ * These are parameters that control the synthesis operation, configurable by the user
+ * from the outside of the app.
+ */
+function parametersFromSynthOptions(synthOptions?: AppSynthOptions) {
+  return synthParametersFromSettings(settingsFromSynthOptions(synthOptions ?? {}));
+}
+
+/**
+ * Temporarily overwrite the `process.env` with a new `env`
+ *
+ * We make the environment immutable in case there are accidental
+ * concurrent accesses.
+ */
+function temporarilyWriteEnv(env: Record<string, string>): () => void {
+  const oldEnv = process.env;
+
+  process.env = immutableEnv({
+    ...process.env,
+    ...env,
+  });
+
+  return () => {
+    process.env = oldEnv;
+  };
+}
```

Comment:
Nit: I think this pattern is actually quite a bite worse than the previous pattern of passing a block. Now you always have to remember to call the cleanup function. Seems rife for errors.